Start of OCS Share API sharees endpoint

rullzer opened the pull request:
Start of implementation of #18117

TODO:
- [ ] Tests, tests, tests and tests
- [ ] Fix pagination (OCS does merging which breaks this maybe later)
- [x] Sort results before splicing
- [ ] Add filtering based on provided share id's (or provided list of users/groups etc??)
- [ ] Did I mention tests?

---
### Todo
- [x] Make share dropdown call the OCS endpoint
- [ ] Add tests for `search()` directly
- [x] Add a filter that prevents sharing with the same user/group/remote again

@DeepDiver1975 FYI

rullzer committed:
Start of OCS Share API sharees endpoint

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/99d3dd21-a620-45e8-a6d3-8af8536bb9d0) was created.

rullzer committed:
Sort sharees

To ensure that pagination is working properly we need to make sure the
shares are always in the same order.

Sorting is first done by label (catches most instances)
If there is a user and a group with the same label we sort by shareType
If there are multiple users with the same label we sort those by
shareWith

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/6b0cf5b5-2d09-4734-9123-1d5a5e535431) was created.

karlitschek commented:
@rullzer Thanks a lot. Looks good so far :-)

rullzer committed:
Empty skeleton for tests

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/49e692ef-1628-4610-9ed7-806421610101) was created.

nickvergessen commented on apps/files_sharing/tests/sharees.php:
```diff
@@ -0,0 +1,50 @@
+<?php
+
+use OCA\Files_Sharing\API\Sharees;
```
missing namespace and copyright

nickvergessen committed:
Move test file to subdir

nickvergessen commented on apps/files_sharing/api/sharees.php:
```diff
@@ -0,0 +1,252 @@
+		$users = [];
+		if ($shareWithGroupOnly) {
+			// Search in all the groups this user is part of
+			$userGroups = $this->groupManager->getUserGroupsIds($this->userSession->getUser());
```
PHP Fatal error:  Call to undefined method Mock_IGroupManager_6a976b4f::getUserGroupsIds() in /home/nickv/ownCloud/master/core/apps/files_sharing/api/sharees.php on line 82

rullzer replied:
Ah crap typo... ```getUserGroupsIds``` -> ```getUserGroupIds```

nickvergessen committed:
Add tests for "getUsers()"

nickvergessen committed:
Add tests for "getGroups()"

nickvergessen committed:
Add tests for "getRemote()"

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/afd8c816-4da6-425d-9857-1ce8dd9db13a) was created.

nickvergessen commented:
@owncloud-bot retest this please

nickvergessen committed:
Use SearchResultSorter

nickvergessen committed:
Call OCS route from share dropdown

nickvergessen committed:
Filter the sharees with the existing shares

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4783cd51-61d7-4f12-adbd-2cc31a197caf) was created.

nickvergessen commented:
@owncloud-bot retest this please

nickvergessen commented:
Replaced by #18234
